be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/DbActor.scala

```diff
@@ -137,157 +180,269 @@ object DbActor extends AskPatternConstants {
-      // open each db associated with each channel name
-      val channelName: String =
-        new String(iterator.next().getKey, StandardCharsets.UTF_8)
-      val channelDb: DB =
-        try { factory.open(new File(channelName), options) }
-        catch {
-          case e: IOException => {
-            logger.error("Could not open channel {}", channelName)
-            throw e
-          }
-
-        }
-
-      // store the channel name and its database in the map
-      initialChannelsMap += (Channel(
-        channelName.toString.stripPrefix(DATABASE_FOLDER)
-      ) -> channelDb)
-    }
-
-    iterator.close()
-    DbActor(mediatorRef, initialChannelsMap.toMap, channelNamesDb, DATABASE_FOLDER, CHANNELS_FOLDER)
+    DbActor(mediatorRef, db, DATABASE_FOLDER)
   }
 
-  sealed case class DbActor(mediatorRef: ActorRef, initialChannelsMap: Map[Channel, DB], channelNamesDb: DB, DATABASE_FOLDER: String, CHANNELS_FOLDER: String) extends Actor with ActorLogging {
-    private val channelsMap: mutable.Map[Channel, DB] = initialChannelsMap.to(collection.mutable.Map)
+  // In the arguments, we might remove the initialChannelsMap and replace the channelNamesDb by the general Db??
+  // FIXME: find way to keep count of channel nb. Idea: store key "channelCount" with value int.getBytes to db and increment it with each channel created
+  sealed case class DbActor(mediatorRef: ActorRef, db: DB, DATABASE_FOLDER: String) extends Actor with ActorLogging {
 
     override def preStart(): Unit = {
-      log.info(
-        s"Actor $self (db) was initialised with a total of ${initialChannelsMap.size} recovered channels"
-      )
-      if (initialChannelsMap.size > DATABASE_MAX_CHANNELS) {
-        log.warning(
-          s"Actor $self (db) has surpassed a large number of active lao channels (${initialChannelsMap.size} > $DATABASE_MAX_CHANNELS)"
-        )
-      }
-
+      //FIXME: check if this should be kept
       super.preStart()
     }
 
-    private def createDatabase(channel: Channel): DB = {
-      val options: Options = new Options()
-      options.createIfMissing(true)
-
-      val channelName: String = s"$DATABASE_FOLDER$channel"
-      val channelDb = factory.open(new File(channelName), options)
-
-      channelNamesDb.put(channelName.getBytes, "".getBytes)
-      channelsMap += (channel -> channelDb)
-
-      channelDb
-    }
-
+    // only for the messages
     private def write(channel: Channel, message: Message): DbActorMessage = {
-      val channelDb: DB = channelsMap.get(channel) match {
-        case Some(channelDb) => channelDb
-        case _ => createDatabase(channel)
+      
+      //encapsulate this inside function
+      Try(db.get(channel.toString.getBytes)) match {
+        case Success(bytes) if bytes != null =>
+          log.info(s"Channel '$channel' already exists.")
+        case _ =>
+          //FIXME: handle type of message (is it actually a problem or is the MessageExample coded in a weird way?)
```

Comment:
If you mean the `util.examples.MessageExample` in the test folder then yes it looks fine to me .

Reply:
Indeed, I have a problem with getting the MessageData out of the Message in order to get the MessageType and haven't been able to solve it yet.